File: Python/stacks_queues.py
### By: Khalil Kum ###

#################### Helper objects/methods ##################
class Stack(object):
  def __init__(self):
    self.data = []

  def isEmpty(self):
    return (len(self.data) == 0)     
  
  def peek(self):
    if self.isEmpty():
      return None
    return self.data[-1]        
 
  def pop(self):
    if self.isEmpty():
      return None
    return self.data.pop()        
      
  def push(self, value):
    self.data.append(value)

def createStack(values):
  stack = Stack()

  for value in values:
    stack.push(value)

  return stack

###################################################################

# 'Balanced Brackets' problem
# Average runtime: O(n)
def isBalanced(brackets):
  if len(brackets) < 2:
    return False

  d = {'}':'{', ']':'[', ')':'('}
  case = "{(["
  temp = []
  for i in brackets:
    if i in case:
      temp.append(i)
    else:
      if  not len(temp):
        return False
      c = temp.pop()
      if c != d[i]:
        return False
  if len(temp)!=0:
      return False
  else:
      return True

# 'Sort a Stack' problem
# Average runtime: O(n^2)
def sortStack(stack):
  if not len(stack.data):
    return stack.data

  temp_stack = Stack()

  while not stack.isEmpty():
    top = stack.pop()

    while (not temp_stack.isEmpty()) and (top > temp_stack.peek()):
      temp_top = temp_stack.pop()
      stack.push(temp_top)

    temp_stack.push(top)

  return temp_stack

# 'A Tale of Two Stacks' problem
# Average runtime: O(n)
class Queue(object):
  def __init__(self):
    self.data_in = []
    self.data_out = []
  
  def peek(self):
    if not self.data_out:
      self.data_out = list(reversed(self.data_in))
      self.data_in = []
    return self.data_out[-1]         
      
  def pop(self):
    head = self.peek()
    del self.data_out[-1]
    return head
      
  def put(self, value):
    self.data_in.append(value)

def runQueries(queries):
  q = Queue()

  if not queries or not len(queries):
    print("INFO: No queries to run")
    exit(1)
  
  for query in queries:
    query = query.split(' ')

    if (len(query) == 2) and (query[0] == '1'):
      q.put(query[1])
    elif (len(query) == 1) and (query[0] == '2'):
      if not q.data_out:
        print("INFO: Nothing to pop()")
      else:
        q.pop()
    elif (len(query) == 1) and (query[0] == '3'):
      print(q.peek())
    else:
      print("ERROR: Invalid query")
    print(f"In-Queue: {q.data_in}")
    print(f"Out-Queue: {q.data_out}")

File: Java/stacks_queues.java
/// By: Merna Alghannam


import java.util.*; 
import java.io.*;

public class StacksAndQueues { 

  static class QueueofS  {  
    static Stack<Character> in = new Stack<Character>();  
    static Stack<Character> out = new Stack<Character>(); 
  
    static void enQueue(char x)  
    {  
        // Move all elements from in to out 
        while (!in.isEmpty()) {  
            out.push(in.pop());   
        }  
  
        // Push item into in  
        in.push(x);  
  
        // Push everything back to in  
        while (!out.isEmpty())  
        {  
            in.push(out.pop());   
        }  
    }  
  
    // Dequeue an item from the queue  
    static char deQueue()  
    {  
        // if first stack is empty  
        if (in.isEmpty())  
        {  
            System.out.println("Q is Empty");  
            System.exit(0);  
        }  
  
        // Return top of in  
        char x = in.peek();  
        in.pop();  
        return x;  
    } 

    static char peak(){
      char x = in.peek(); 
      return x;
    } 

    static void printStack(Stack<Character> s) { 
      // If stack is empty then return 
      if (s.isEmpty()) 
          return; 
      
      char x = s.peek(); 
  
      // Pop the top element of the stack 
      s.pop(); 
  
      // Recursively call the function PrintStack 
      printStack(s); 
  
      // Print the stack element starting 
      // from the bottom 
      System.out.print(x + " "); 
  
      // Push the same element onto the stack 
      // to preserve the order 
      s.push(x); 
    } 

    static void printQueue(){
      // Move all elements from in to out 
      while (!in.isEmpty()) {  
          out.push(in.pop());  
      } 

      printStack(out);

      // Push everything back to in  
      while (!out.isEmpty()) {  
          in.push(out.pop()); 
      } 
      System.out.println();
    }
  };
  
  // function to check if brackets are balanced 
  static boolean areBracketsBalanced(String str) { 
      Stack<Character> s = new Stack<Character>(); 

      for (int i = 0; i < str.length(); i++)  
      { 
          char x = str.charAt(i); 

          if (x == '(' || x == '[' || x == '{')  { 
              // Push the element in the stack 
              s.push(x); 
              continue; 
          } 

          // IF current current character is not opening 
          // bracket, then it must be closing. So stack 
          // cannot be empty at this point. 
          if (s.isEmpty()) 
              return false; 
          char c; 
          switch (x) { 
          case ')': 
              c = s.pop(); 
              if (c == '{' || c == '[') 
                  return false; 
              break; 
          case '}': 
              c = s.pop(); 
              if (c == '(' || c == '[') 
                  return false; 
              break; 
          case ']': 
              c = s.pop(); 
              if (c == '(' || c == '{') 
                  return false; 
              break; 
          } 
      } 

      // Check Empty Stack 
      return (s.isEmpty()); 
  } 

  // This function return the sorted stack 
  public static Stack<Integer> sortstack(Stack<Integer> original) { 
      Stack<Integer> newStack = new Stack<Integer>(); 
      while(!original.isEmpty()) 
      { 
          // pop out the first element 
          int tmp = original.pop(); 
        
          // while temporary stack is not empty and 
          // top of stack is greater than temp 
          while(!newStack.isEmpty() && newStack.peek() > tmp) { 
              // pop from temporary stack to original stack 
            original.push(newStack.pop()); 
          } 
            
          // push temp in tempory of stack 
          newStack.push(tmp); 
      } 
      return newStack; 
  } 

  public static void runQueries(String queries){
    QueueofS q = new QueueofS();  

    if (queries == null || queries.length() == 0){
      System.out.println("INFO: No queries to run");
      return;
    }

    String[] query = queries.split(" ");
    
    for (int i = 0; i < query.length; i++){
      if (query[i].length() == 2 && query[i].charAt(0) == '1'){     
        q.enQueue(query[i].charAt(1));
      }else if (query[i].length() == 1 && query[i].charAt(0) == '2'){
        q.deQueue();
      }else if (query[i].length() == 1 && query[i].charAt(0) == '3'){
        System.out.println("peaking element at top: " + q.peak());
      }else{
        System.out.println("ERROR: Invalid query");
      }
      System.out.println("printing queue:");
      q.printQueue();
    } 
  }
  
  // Driver code 
  public static void main(String[] args) { 
      String expr = "([{}])"; 

      // Function call  for areBracketsBalanced
      if (areBracketsBalanced(expr)) 
          System.out.println("Balanced "); 
      else
          System.out.println("Not Balanced "); 

      // Function call for sortstack
      Stack<Integer> input = new Stack<Integer>(); 
      input.add(34); 
      input.add(3); 
      input.add(31); 
      input.add(98); 
      input.add(92); 
      input.add(23); 
  
      Stack<Integer> tmpStack = sortstack(input); 
      System.out.println("Sorted numbers are:"); 
    
      while (!tmpStack.empty()) 
      { 
          System.out.print(tmpStack.pop()+" "); 
      }  
      System.out.println("\n");

      //function call for runQueries
      runQueries("12 23 11 3 22 33 1 2 3");
  } 
}
